toplevel/topdirs.ml

```diff
@@ -545,15 +545,66 @@ let () =
   reg_show_prim "show_type"
     (fun env loc id lid ->
        let _path, desc = Env.lookup_type ~loc lid env in
-       [ Sig_type (id, desc, Trec_not, Exported) ]
+       [ Sig_type (id, desc, Trec_first, Exported) ]
     )
     "Print the signature of the corresponding type constructor."
 
+(* Each registered show_prim function is called in turn
+ * and any output produced is sent to std_out.
+ * Two show_prim functions are needed for constructors,
+ * one for exception constructors and another for
+ * non-exception constructors (normal and extensible variants). *)
+let is_exception_constructor env type_expr =
+  Ctype.equal env true [type_expr] [Predef.type_exn]
+
+let is_extension_constructor = function
+  | Cstr_extension _ -> true
+  | _ -> false
+
+let () =
+  (* This show_prim function will only show constructor types
+   * that are not also exception types. *)
+  reg_show_prim "show_constructor"
+    (fun env loc id lid ->
+       let desc = Env.lookup_constructor ~loc Env.Positive lid env in
+       if is_exception_constructor env desc.cstr_res then
+         raise Not_found;
+       let path =
+         match Ctype.repr desc.cstr_res with
+         | {desc=Tconstr(path, _, _)} -> path
+         | _ -> raise Not_found
+       in
+       if is_extension_constructor desc.cstr_tag then
+         let ret_type =
+           if desc.cstr_generalized then Some Predef.type_exn
```

Comment:
I don't think that this is right -- you use `Predef.type_exn` for all extension constructors. You can (you should) test with

```
# type 'a t = ..;;
# type _ t += A : int t;;
[%%expect{|
type 'a t = ..
type _ t += A : int t
|}];;
# #show A;;
[%%expect{|
|}];;
```

I'm not completely sure but I think that `desc.cstr_res` should be usable here.

Reply:
i think you are correct, cstr_res is the type_expr of the result if I understand correctly, also I just noticed I need to be passing in the correct type parameters too, which live on the type_declaration - and with that we get:
```
# type 'a t = ..;;
# type _ t += A : int t;;
# #show A;;
type 'a t += A : int t
```
which seems right